src/test/java/com/wikia/webdriver/testcases/adstests/TestAdsTrackingOptInRequestsF2.java

```diff
@@ -0,0 +1,96 @@
+import com.wikia.webdriver.pageobjectsfactory.pageobject.adsbase.AdsFandomObject;
+import org.openqa.selenium.TimeoutException;
+import org.testng.annotations.Test;
+
+import java.util.concurrent.TimeUnit;
+
+public class TestAdsTrackingOptInRequestsF2 extends AdsFandomTestTemplate {
+
+  @NetworkTrafficDump(useMITM = true)
+  @Execute(trackingOptIn = false)
+  @Test(groups = "AdsOptInModalF2")
+  public void testModalVisible() {
+    networkTrafficInterceptor.startIntercepting();
+    AdsFandomObject page = loadPage(FandomAdsDataProvider.PAGE_HIVI_UAP_ARTICLE);
+    TrackingOptInModal modal = new TrackingOptInModal(page);
+
+    Assertion.assertTrue(modal.isVisible());
+  }
+
+  @NetworkTrafficDump(useMITM = true)
+  @Test(groups = "AdsOptInModalF2")
+  public void testNoRequestForAdBeforeModal() throws InterruptedException {
+    networkTrafficInterceptor.startIntercepting();
+    AdsFandomObject fandomPage = loadPage(FandomAdsDataProvider.PAGE_HIVI_UAP_ARTICLE);
+    TrackingOptInModal modal = new TrackingOptInModal(fandomPage);
+    fandomPage.waitForPageLoad();
+
+    TimeUnit.SECONDS.sleep(10);
+    Assertion.assertFalse(fandomPage.wasRequestForAdSend(networkTrafficInterceptor));
+    modal.clickAcceptButton();
+
+    TimeUnit.SECONDS.sleep(5);
+    Assertion.assertTrue(fandomPage.wasRequestForAdSend(networkTrafficInterceptor));
+  }
+
+  @NetworkTrafficDump(useMITM = true)
+  @Execute(trackingOptIn = true)
+  @Test(
+          dataProviderClass = TrackingOptInDataProvider.class,
+          dataProvider = "f2TrackingURLsForOptIn",
+          groups = "AdsOptInAcceptF2"
+  )
+  public void testRequestsForOptIn(String pattern, boolean isExpected) {
+    testRequests(pattern, isExpected);
+  }
+
+  @NetworkTrafficDump(useMITM = true)
+  @Execute(trackingOptIn = false, trackingOptOut = true)
+  @Test(
+          dataProviderClass = TrackingOptInDataProvider.class,
+          dataProvider = "f2TrackingURLsForOptOut",
+          groups = "AdsOptInRejectF2"
+  )
+  public void testRequestsForOptOut(String pattern, boolean isExpected) {
+    testRequests(pattern, isExpected);
+  }
+
+  private void testRequests(String pattern, boolean isExpected) {
+    AdsFandomObject page = loadPageWithGeo();
+
+    if (isExpected) {
+      Assertion.assertTrue(isRequestFound(pattern, page));
+    } else {
+      Assertion.assertFalse(isRequestFound(pattern, page));
+    }
+  }
+
+  private AdsFandomObject loadPageWithGeo() {
+    AdsFandomObject page = loadPage(FandomAdsDataProvider.PAGE_HIVI_UAP_ARTICLE);
+    TrackingOptInPage.setGeoCookie(driver, "EU", "PL");
+    networkTrafficInterceptor.startIntercepting();
+    page.refreshPage();
+    page.waitForPageLoad();
+    return page;
+  }
+
+  private boolean isRequestFound(String pattern, AdsFandomObject page) {
+    try {
+      page.wait.forSuccessfulResponseByUrlPattern(networkTrafficInterceptor, pattern);
+    } catch (TimeoutException error) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Either log or rethrow this exception. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar-s4/coding_rules#rule_key=squid%3AS1166)